### ISSUE TITLE
Names do matter where they should not (broken alpha conversion)

### ISSUE THREAD

**Issue by nicolabotta** [idris-lang/Idris-dev#3449]
The program
```
> import Syntax.PreorderReasoning

> %default total

> rev' : (xs : List a) -> (ys : List a) -> List a
> rev' acc []        = acc
> rev' acc (x :: xs) = rev' (x :: acc) xs 

> rev : List a -> List a
> rev = rev' [] 

> lemma : (xs, ys : List a) ->  rev' xs ys  = rev ys ++ xs 
> lemma acc [] = Refl
> lemma acc (x :: xs) = ( rev' acc (x :: xs) )
>                     ={ lemma (x :: acc) xs }= 
>                       ( rev xs ++ ([x] ++ acc) ) 
>                     ={ appendAssociative (rev xs) [x] acc }=  
>                       ( (rev xs ++ [x]) ++ acc )
>                     ={ cong {f = \ X => X ++ acc} (sym (lemma [x] xs)) }=
>                       ( rev (x :: xs) ++ acc )
>                     QED 
```
type checks fine. But replacing
```
cong {f = \ X => X ++ acc} (sym (lemma [x] xs))
```
with
```
cong {f = \ Z => Z ++ acc} (sym (lemma [x] xs))
```
yields
```
- + Errors (1)
 `-- namesMatter.lidr line 19 col 55:
     When checking right hand side of lemma with expected type
             rev' acc (x :: xs) = rev (x :: xs) ++ acc
     
     When checking an application of function sym:
             Type mismatch between
                     rev' xs ys = rev ys ++ xs (Type of lemma xs ys)
             and
                     b = acc (Expected type)
             
             Specifically:
                     Type mismatch between
                             List a
                     and
                             Nat
```
! This is with Idris 0.12.2-git:1681ef9.

**Comment by xekoukou** [comment]
The problem is with the letter 'Z' specifically, not any other letter.

**Comment by ahmadsalim** [comment]
@nicolabotta I think this is because you are running into the case of pattern lambdas. So `\Z => ...` actually creates an anonymous pattern matching lambda `\x => case x of Z => ...`.

The way you can get around this is to use lower-cased variable names in your lambda-expressions which should be stable.

**Comment by ahmadsalim** [comment]
I would guess it's hard in general to do this distinction beforehand, and have good inference. @idris-lang/commiters Maybe we should enforce the constructor name/upper-case requirement soon? I think it would be easier to provide a good error message in this case.

**Comment by xekoukou** [comment]
@nicolabotta Yes, it is true. I had similar problems with Z, but it happens for all names of a variable that match the name of a constructor.

```
data Test = Construct
         
id : Nat -> Nat
id Construct = Construct
```

```
When checking an application of Main.id:
        Type mismatch between
                Test (Type of Construct)
        and
                Nat (Expected type)

```

**Comment by nicolabotta** [comment]
@ahmadsalim Thanks for the clarification! I am not sure whether this should be the expected behavior though, essentially for two resons:

1) I think that, in principle, adding an import directive to a program should in the worst case lead to `Can't disambiguate` errors, not to errors like above.

2) The fact that `cong {f = \ X => X ++ acc} (sym (lemma [x] xs))` cannot be replaced by `cong {f = \ Z => Z ++ acc} (sym (lemma [x] xs))` seem to violate the rule that local names (in fact, even implicit ones) bind more strongly than global names.

I am not sure that I fully understand Idris qualification rules but, in this specific example, I would expect it to be perfectly legitimate to replace
```
cong {f = \ X => X ++ acc} (sym (lemma [x] xs))
```
with
```
cong {f = \ Z => Z ++ acc} (sym (lemma [x] xs))
```
Moreover, I would expect that, if one would like some of the `Z`s in the above expression to refer to specific data constructors (there might be many data constructors in the global scope denoted by `Z`) one would have to fully qualify those `Z`s, like for instance in
```
cong {f = \ Z => Prelude.Nat.Z ++ Z} (sym (lemma [x] xs))
```

**Comment by melted** [comment]
But there is no ambiguity if the only Z in scope is the Nat constructor

**Comment by nicolabotta** [comment]
@melted Of course there is an ambiguity: we have the local name `Z` introduced by the `\ Z =>` clause and we have the global name `Prelude.Nat.Z`. As Idris requires one to write
```
> f : (xs : List a) -> map Prelude.Basics.id xs = xs
```
if `id` is meant to represent the identity function and not a local implicit variable, I would expect Idris to require one to write
```
> f : Nat -> Nat
> f = \ Z => Prelude.Nat.Z
```
if `f` is not meant to implement the identity function. Conversely, I would expect Idris to be happy with either
```
> f : Nat -> Nat
> f = \ Z => Z
```
or
```
> f : Nat -> Nat
> f = \ LALA => LALA
```
implementations of the identity function for `Nat`. In other words, I think that it is a mistake to make the semantics of 
```
> f : Nat -> Nat
> f = \ LALA => LALA
```
depend on the context. Users might import modules in which `LALA` is the name of a data constructor. They might end up with implementations of  `f` that type check fine but do something completely different from what they expect.

**Comment by ahmadsalim** [comment]
@nicolabotta I think the issue here is that one can use lambda-expressions both for pattern-matching and for simple abstraction, and there is currently way no way to distinguish between what the user wants (which we all agree on). 

A current workaround is to only use lower-case names for variables and upper-case for constructors.

In a long term I see two possible solution:

* One is to enforce this distinction in naming, so one can only name constructors with capitals and variables with lower-case

* The other is to distinguish the pattern matching lambdas from the ordinary ones, e.g. by using `\case` like in Haskell for pattern matching ones.

**Comment by xekoukou** [comment]
According to Edwin's [2013 paper] (https://eb.host.cs.st-andrews.ac.uk/drafts/impldtp.pdf), this is the order in which elaboration works (page 15).

Specifically for my example in which a type is provided, Constructors which do not give the correct Type should be dismissed and be considered as simple variable names.

**Comment by nicolabotta** [comment]
@ahmadsalim I have no problems using lower-case names in the example above. On the other hand, there are situations in which one would like to use upper-case names for variables in anonymous functions. Generally speaking, I believe that distinctions based on upper/lower-case rules should be avoided  whenever possible and that using `\case` to distinguish between pattern matching lambdas and ordinary expressions would be a very clean solution.
